types/time.go

```diff
@@ -1563,234 +1607,66 @@ func extractSingleTimeValue(unit string, format string) (int64, int64, int64, fl
-	fields := strings.Split(format, ":")
-	if len(fields) != 3 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	hours, err := strconv.ParseFloat(fields[0], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	minutes, err := strconv.ParseFloat(fields[1], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	_, _, _, value, err := extractSecondMicrosecond(fields[2])
-	if err != nil {
-		return 0, 0, 0, 0, errors.Trace(err)
-	}
-
-	return 0, 0, 0, hours*float64(gotime.Hour) + minutes*float64(gotime.Minute) + value, nil
-}
-
-// extractHourSecond extracts hour and second from a string and its format is `HH:MM:SS`.
-func extractHourSecond(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, ":")
-	if len(fields) != 3 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	hours, err := strconv.ParseFloat(fields[0], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	minutes, err := strconv.ParseFloat(fields[1], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	seconds, err := strconv.ParseFloat(fields[2], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	return 0, 0, 0, hours*float64(gotime.Hour) + minutes*float64(gotime.Minute) + seconds*float64(gotime.Second), nil
-}
-
-// extractHourMinute extracts hour and minute from a string and its format is `HH:MM`.
-func extractHourMinute(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, ":")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	hours, err := strconv.ParseFloat(fields[0], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	minutes, err := strconv.ParseFloat(fields[1], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	return 0, 0, 0, hours*float64(gotime.Hour) + minutes*float64(gotime.Minute), nil
-}
-
-// extractDayMicrosecond extracts day and microsecond from a string and its format is `DD HH:MM:SS.FFFFFF`.
-func extractDayMicrosecond(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, " ")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+// extractTimeValue extracts years, months, days, microseconds from a string
+// MySQL permits any punctuation delimiter in the expr format.
+// See https://dev.mysql.com/doc/refman/8.0/en/expressions.html#temporal-intervals
+func extractTimeValue(format string, index, cnt int) (int64, int64, int64, float64, error) {
+	neg := false
+	originalFmt := format
+	if len(format) > 0 && format[0] == '-' {
```

Comment:
How about trim prefix of this string, `"- 1:1"` and `" - 1:1"` will be both recognized as negative in MySQL.

Reply:
add trim space
PTAL @qw4990